pkg/pillar/portprober/portprober.go

```diff
@@ -0,0 +1,956 @@
+// stack (currently only Linux networking is supported by EVE).
+// A mock implementation is also provided for unit testing purposes.
+type ReachabilityProber interface {
+	// Probe reachability of <dstAddr> via the given port.
+	// If <dstAddr> is IP address and not hostname, <dnsServers> can be nil.
+	Probe(ctx context.Context, portIfName string, srcIP net.IP, dstAddr net.Addr,
+		dnsServers []net.IP) error
+}
+
+// ProbeStatus is published whenever the selected port for a given multipath route
+// changes.
+type ProbeStatus struct {
+	NetworkInstance uuid.UUID
+	MPRoute         types.IPRouteConfig
+	SelectedPortLL  string
+	SelectedAt      time.Time
+}
+
+// multipathRoute is used only internally to store configs of multipath routes and
+// labels of their currently selected output ports.
+type multipathRoute struct {
+	ProbeStatus
+	NIPortLabel string
+}
+
+// portStatus - used only internally to track the probing state of every probed port.
+type portStatus struct {
+	logicallabel string
+	sharedlabels []string
+	ifName       string
+	cost         uint8
+	isWwan       bool
+	localAddrs   []net.IP
+	nextHops     []net.IP
+	dnsServers   []net.IP
+	newlyAdded   bool
+	nhProbe      probeStatus
+	userProbes   map[types.ConnectivityProbe]*probeStatus
+}
+
+func (ps portStatus) matchesLabels(labels ...string) bool {
+	for _, label := range labels {
+		if ps.logicallabel != label && !generics.ContainsItem(ps.sharedlabels, label) {
+			return false
+		}
+	}
+	return true
+}
+
+func (ps portStatus) upCount(probeCfg types.NIPortProbe) int {
+	var upCount int
+	if probeCfg.EnabledGwPing && ps.cost <= probeCfg.GwPingMaxCost {
+		if ps.nhProbe.connIsUP {
+			upCount++
+		}
+	}
+	userProbe := ps.userProbes[probeCfg.UserDefinedProbe]
+	if userProbe != nil {
+		if userProbe.connIsUP {
+			upCount++
+		}
+	}
+	return upCount
+}
+
+type probeStatus struct {
+	refCount   int
+	connIsUP   bool
+	failedCnt  uint32 // continuous fail count, reset on success
+	successCnt uint32 // continuous success count, reset on fail
+	avgLatency time.Duration
+}
+
+// NewPortProber is a constructor for PortProber.
+func NewPortProber(log *base.LogObject, config Config,
+	reachProberICMP, reachProberTCP ReachabilityProber) *PortProber {
+	if config.NHToUserProbeRatio < minNHToUserProbeRatio {
+		config.NHToUserProbeRatio = minNHToUserProbeRatio
+	}
+	prober := &PortProber{
+		log:             log,
+		config:          config,
+		reachProberICMP: reachProberICMP,
+		reachProberTCP:  reachProberTCP,
+		routes:          make(map[string]*multipathRoute),
+		ports:           make(map[string]*portStatus),
+	}
+	prober.probeTicker = time.NewTicker(config.NHProbeInterval)
+	go prober.runProbing()
+	return prober
+}
+
+// StartPortProbing tells PortProber to start periodic probing of network ports
+// for this multipath route.
+// It is called by zedrouter whenever a new multipath route is configured for an existing
```

Comment:
Is "multipath route" the corrrect term to use here?

A linux multipath route is a route for one destination prefix which has two or more nexthops (interface,gateway). But isn't the concept here a NI with multiple exits?

Thus if I somehow configure a route for 10.1.2.3 with nexthops 11.0.0.1 on eth0 and 12.0.0.1 on eth1, would there be an new prober created? 
Note that we might not be able to configure such routes today, but I concerned about terminology confusion today and tomorrow.

Reply:
I think the api today does allow such a case for loadsharing ip static routes. so in this function, we need to detect this route, regardless of gw, has already been probed

Reply:
So the idea behind the API enhancements done in https://github.com/lf-edge/eve-api/pull/53 is to extend/generalize the Local NI network model.
First, as you pointed out, we want to allow NI having multiple ports (exits to external networks).
Second, to solve the problem of the default route, which with multiple ports is often a multipath route but we need to pick one port at a time, I decided to generalize the probing and port selection functionality to any static IP route with multiple exits (instead of creating solution for the default route only). Today user configures dst network + gateway IP for static route. After those changes, they will be able to select output port(s) additionally to or instead of gateway IP using logical/shared port labels (see [here](https://github.com/lf-edge/eve-api/pull/53/files#diff-445699b886092aa0f66615f9134a9a1ade8c21fbc992adf2f8f2bf1855eaf675R89-R103)). If there are multiple valid output ports for a route, user can select them using a shared label. Then we get a multipath route. Probing will be used to select one output port a time (and user can customize method & endpoint to use for probing). I believe that users will prefer to configure dst+port without gateway IP and let EVE to find out what the gateway IP for a particular port is (based on the port config received from DHCP or applied statically).
Still they can configure all 3 parameters: dst + port + gateway. But then only ports whose subnets match the gateway IP will be considered. In normal scenarios this prevents multipath routes (hence with the current API we cannot really create them).
With this enhanced model, the default route will be treated as yet another multipath IP route (either user defined or automatically generated by EVE based on the NI config).

Reply:
Does this we'll use the Linux kernel multipath route support? Or you will be handling the path selection using the probing in user space?

Reply:
> Or you will be handling the path selection using the probing in user space

This. We will be always installing only normal routes with single gateway + port. In zedrouter/portprober the port selection will be made (and possibly changed in runtime to fail-over when probing through the current port starts to fail).